Sources/SwiftGRPCNIO/GRPCClientChannelHandler.swift

```diff
@@ -0,0 +1,200 @@
+    initialMetadataPromise: EventLoopPromise<HTTPHeaders>,
+    statusPromise: EventLoopPromise<GRPCStatus>,
+    responseObserver: ResponseObserver<ResponseMessage>
+  ) {
+    self.initialMetadataPromise = initialMetadataPromise
+    self.statusPromise = statusPromise
+    self.responseObserver = responseObserver
+  }
+
+  /// Observe the given status.
+  ///
+  /// The `status` promise is __succeeded__ with the given status despite `GRPCStatus` being an
+  /// `Error`. If `status.code != .ok` then the initial metadata and response promises (if applicable)
+  /// are failed with the given status.
+  ///
+  /// - Parameter status: the status to observe.
+  internal func observeStatus(_ status: GRPCStatus) {
+    if status.code != .ok {
+      self.initialMetadataPromise.fail(error: status)
+      self.responsePromise?.fail(error: status)
+    }
+    self.statusPromise.succeed(result: status)
+  }
+}
+
+extension GRPCClientChannelHandler: ChannelInboundHandler {
+  public typealias InboundIn = GRPCClientResponsePart<ResponseMessage>
+
+  /// Reads inbound data.
+  ///
+  /// On receipt of:
+  /// - headers: the initial metadata promise is succeeded.
+  /// - message: the message observer is called with the message; for unary responses a response
+  ///   promise is succeeded, otherwise a callback is called.
+  /// - status: the status promise is succeeded; if the status is not `ok` then the initial metadata
+  ///   and response promise (if available) are failed with the status. The channel is then closed.
+  public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    guard self.inboundState != .ignore else { return }
+
+    switch unwrapInboundIn(data) {
+    case .headers(let headers):
+      guard self.inboundState == .expectingHeaders else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.initialMetadataPromise.succeed(result: headers)
+      self.inboundState = .expectingMessageOrStatus
+
+    case .message(let message):
+      guard self.inboundState == .expectingMessageOrStatus else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.responseObserver.observe(message)
+
+    case .status(let status):
+      guard self.inboundState == .expectingMessageOrStatus else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.observeStatus(status)
+
+      // We don't expect any more requests/responses beyond this point.
+      self.close(ctx: ctx, mode: .all, promise: nil)
+    }
+  }
+}
+
+extension GRPCClientChannelHandler: ChannelOutboundHandler {
+  public typealias OutboundIn = GRPCClientRequestPart<RequestMessage>
+  public typealias OutboundOut = GRPCClientRequestPart<RequestMessage>
+
+  public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    guard self.inboundState != .ignore else { return }
```

Comment:
Did you really mean `.inboundState` here?

Reply:
Nope!